tests/abi/runtime_abi_analysis.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * This file is distributed under the MIT License. See LICENSE.md for details.
+ */
+
+#include <setjmp.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "for_each.h"
+#include "functions.h"
+
+#ifdef TARGET_x86_64
+#include "platform_specific/x86_64.h"
+#elif TARGET_x86
+#include "platform_specific/x86.h"
+#else
+#include "platform_specific/x86_64.h"
+static_assert(0, "unsupported platform architecture");
+#endif
+
+/* LFSR generator */
+
+static uint64_t lfsr = 0xdc984ef2647af904ull;
+uint64_t get_next_lfsr(void) {
+  for (uint32_t iteration = 0; iteration < 64; ++iteration)
+    lfsr = (lfsr >> 1) | ((((lfsr >> 1) ^ (lfsr >> 2)) & 1) << 63);
+  return lfsr;
+}
+
+/* Helpers */
+
+#define GET_1ST_HELPER(THIS_ONE, ...) THIS_ONE
+#define GET_1ST(THIS_ONE, ...) GET_1ST_HELPER(THIS_ONE, __VA_ARGS__)
+#define TMP_REGISTER GET_1ST(REGISTER_LIST)
+
+#define STACK_VALUE_COUNT 32
+#define REGISTER_COUNT COUNT_ARGUMENTS(REGISTER_LIST)
+#define GENERATED_COUNT REGISTER_COUNT + STACK_VALUE_COUNT
+
+#ifdef STATE_LOCATION
+#undef STATE_LOCATION
+#endif
+#define STATE_LOCATION "randomized_state"
+
+char *register_list[REGISTER_COUNT] = {
+  COMMA_SEPARATED_FOR_EACH(STRINGIFY, REGISTER_LIST)
+};
+
+static __attribute((used)) register_type randomized_state[GENERATED_COUNT];
+void regenerate_state(void) {
+  for (uint32_t counter = 0; counter < GENERATED_COUNT; ++counter)
+    randomized_state[counter] = get_next_lfsr();
+}
+void print_current_state(void) {
+  /* Dump Stack */
+  puts("- Stack:");
+  for (uint32_t index = REGISTER_COUNT; index < GENERATED_COUNT; ++index) {
+    printf("    - Offset: 0x%02x\n      Value: 0x",
+           (index - REGISTER_COUNT) * REGISTER_SIZE);
+    PRINT_VALUE(uint64_t, randomized_state[index]);
+    puts("");
```

Comment:
What's the advantage of printing it at 64-bit shots compared to a by array?

Reply:
I think it'd be simpler.
But I don't want to make this longer than necessary. Maybe put a TODO. Also, please document in `README.md` where offset `0` is. From a callee's perspective, is it the integer right after the saved return address?

Reply:
So it's pretty much
```
push rcx // offset 16
push rbx // offset 8
push rax // offset 0
call my_function
```
There's no easy way to find out whether there is an address on the stack, shadow space for something or anything else - so I didn't bother. That represents `push`es right before the call instruction.

Reply:
Uhm, if that's exactly the code you have, you know 100% sure that the offset 0 integer is going to be right after the saved return address pushed by the `call` instruction.
If this is the case, what I suggested before should hold, right? Doc doc doc.

Reply:
I see two options: either `call` pushes the return address on the stack or nothing is pushed.
In both cases we can say that offset `0` corresponds to the value of the stack pointer *right before* the `call` instruction. Is this right?

> I'm trying to make as few assumptions as I can

Making as few assumptions as possible is important, but it cannot become a reason not to document and provide sane expectation to the next dev that will look at this code `:)`.

Reply:
Aaah. I missed that part.
I think that if there's anything in there the test is f-'d.
I think we either find a way to `assert` things are the way we expect, or we should just assume it and document things accordingly.

Reply:
Yeah, pretty much. And that's what makes me so sad about the situation.

Reply:
Wait, why do you say that you cannot control the call instruction?

I see [the following](https://github.com/revng/revng-qa/pull/4/files#diff-294dd971aa761265be5b8dcc15dfaad6bee2d5ddb4413d015a1157e61767d66cR99):

```

    __asm__ (                                                                 \
      /* Popularize the stack with random data */                             \
      REPEAT(PUSH_CONSTANT, STACK_VALUE_COUNT, GENERATED_COUNT, TMP_REGISTER) \
      /* Popularize all the registers with random data */                     \
      INDEXED_FOR_EACH(FILL_REGISTER, REGISTER_LIST)                          \
                                                                              \
      /* Call the function */                                                 \
      CALL_A_FUNCTION(FUNCTION_NAME)                                          \
                                                                              \
      /* Restore the stack */                                                 \
      RESTORE_STACK(STACK_VALUE_COUNT)                                        \
    );                                                                        \
```

We control the call!

Reply:
We control what happens BEFORE the `call`. And what happens AFTER the `call`. What I'm scared of is what the call instruction itself does. Sure, it's pretty simple on most common architectures. But some exotic ones that do some weird stuff probably exist.

Reply:
I like to make statements that can be either true or false, no matter ABI, architecture or whatever.
Here's my statement:
> Offset 0 corresponds to the value of the stack pointer right before the call instruction.

True or false? If it's true, I think it's an important piece of information that we should put down as a definition of what the `Stack` data in the YAML is.

Reply:
I'd phrase it like

> Offset `0` indicates the value corresponding to the last unpopped register-sized value on the stack right before the call instruction. Offset `8/4/RegisterSize` - second to last unpopped value. Offset 2*`8/4/RegisterSize` - third and so on.

Does that work?

Reply:
Not really, that's very long and convoluted.
Can you make an example that falsifies the one above?

Reply:
If you say "unpopped" you're forcing the user to know that something has been pushed, which is an implementation detail to me. There could be an implementation with no-push and no-pops.
If you want, you can add "assuming the stack grows towards low addresses, ...". But other than that I'd try to explain what that data is in a simple way, leaving out implementation details.